File: python/data_structures/stack/stack.py
"""Stack implementation in Python."""


class Stack:
    """Stack implements the stack ADT."""

    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if self.size():
            return self.items.pop()
        else:
            return None

    def peek(self):
        if self.size():
            return self.items[self.size() - 1]
        else:
            return None

    def is_empty(self):
        return self.size() == 0

    def size(self):
        return len(self.items)

    def __str__(self):
        output = "-" * 80
        output += "\nStack:"
        if self.size():
            for i in range(self.size()):
                output += f"\n\t{i + 1:2}:{self.items[i]}"
            output += f"\nSize: {self.size()}"
        else:
            output += " [] (emtpy)"

        return output


if __name__ == "__main__":
    s = Stack()
    print("Is empty:", s.is_empty())
    print(s)

    s.push("Neko")
    s.push(7)
    s.push("Inu")

    print(s)
    print("Peek:", s.peek())

    last = s.pop()
    print("Poped:", last)
    print(s, f"(size: {s.size()})")

File: go/pkg/stack/stack.go
package stack

import "fmt"

type Stack struct {
	items []any
}

func New() *Stack {
	return &Stack{}
}

func (s *Stack) Push(item any) {
	s.items = append(s.items, item)
}

func (s *Stack) Pop() any {
	if s.Size() > 0 {
		last := s.items[len(s.items)-1]
		s.items = s.items[:len(s.items)-1]
		return last
	}
	return nil
}

func (s *Stack) Size() int {
	return len(s.items)
}

func (s *Stack) Peek() any {
	return s.items[len(s.items)-1]
}

func (s *Stack) IsEmpty() bool {
	return s.Size() == 0
}

func PrintStack(s *Stack) {
	if s.IsEmpty() {
		println("Stack: empty.")
	} else {
		outputStr := "--------------------------------------------------------------------------------\nStack:"
		for i, v := range s.items {
			outputStr += fmt.Sprintf("\n\t%2d: %v", i+1, v)
		}
		outputStr += fmt.Sprintf("\nSize: %d", s.Size())
		println(outputStr)
	}
}
